src/Toml/Type/AnyValue.hs

```diff
@@ -38,41 +53,54 @@ instance Eq AnyValue where
 instance NFData AnyValue where
     rnf (AnyValue val) = rnf val
 
+-- Value mismatch error
+data MatchError = MatchError
+    { valueExpected :: TValue
+    , valueActual   :: AnyValue
+    } deriving (Eq, Show, Generic, NFData)
+
 ----------------------------------------------------------------------------
 -- Matching functions for values
 ----------------------------------------------------------------------------
 
 -- | Extract 'Bool' from 'Value'.
-matchBool :: Value t -> Maybe Bool
-matchBool (Bool b) = Just b
-matchBool _        = Nothing
+matchBool :: Value t -> Either MatchError Bool
+matchBool (Bool b) = Right b
+matchBool value    = leftMatchError value
 
 -- | Extract 'Integer' from 'Value'.
-matchInteger :: Value t -> Maybe Integer
-matchInteger (Integer n) = Just n
-matchInteger _           = Nothing
+matchInteger :: Value t -> Either MatchError Integer
+matchInteger (Integer n) = Right n
+matchInteger value       = leftMatchError value
 
 -- | Extract 'Double' from 'Value'.
-matchDouble :: Value t -> Maybe Double
-matchDouble (Double f) = Just f
-matchDouble _          = Nothing
+matchDouble :: Value t -> Either MatchError Double
+matchDouble (Double f) = Right f
+matchDouble value      = leftMatchError value
 
 -- | Extract 'Text' from 'Value'.
-matchText :: Value t -> Maybe Text
-matchText (Text s) = Just s
-matchText _        = Nothing
+matchText :: Value t -> Either MatchError Text
+matchText (Text s) = Right s
+matchText value    = leftMatchError value
 
 -- | Extract 'DateTime' from 'Value'.
-matchDate :: Value t -> Maybe DateTime
-matchDate (Date d) = Just d
-matchDate _        = Nothing
+matchDate :: Value t -> Either MatchError DateTime
+matchDate (Date d) = Right d
+matchDate value    = leftMatchError value
 
 -- | Extract list of elements of type @a@ from array.
-matchArray :: (AnyValue -> Maybe a) -> Value t -> Maybe [a]
+matchArray :: (AnyValue -> Either MatchError a) -> Value t -> Either MatchError [a]
 matchArray matchValue (Array a) = mapM (liftMatch matchValue) a
-matchArray _            _       = Nothing
+matchArray _          value     = leftMatchError value
+
+-- | Extract NonEmpty list of elements of type @a@ from array.
+matchArrayNE :: (AnyValue -> Either MatchError a) -> Value t -> Either MatchError (NE.NonEmpty a)
```

Comment:
This function shouldn't be in `Type` directory. This module only for basic and primitive stuff. It's better to handle `NonEmpty` error in `BiMap` module.